avfilter: New scene classification filter

jailuthra opened the pull request:
**Why?**
https://github.com/livepeer/lpms/issues/248

**What?**
- Replace our old `vf_lvpdnn` filter with `vf_livepeer_dnn` which is simpler, and uses the newer APIs in ffmpeg v4.4

**TODO**
- [x] Update documentation
- [x] Remove old pre-loading API from avfilter
- [x] Use newer session config API in v4.4 instead of our custom logic

oscar-davids committed:
libavfilter: add lvpdnn filter

jailuthra committed:
avfilter/vf_livepeer_dnn: Add scene classification filter

Works combined with hwdownload,format=nv12 if NVDEC is used.

jailuthra committed:
avfilter/vf_lvpdnn: Disable filter

jailuthra commented on libavfilter/avfilter.h:
```diff
@@ -1171,6 +1171,22 @@ char *avfilter_graph_dump(AVFilterGraph *graph, const char *options);
+ * Free all buffer for lvpdnn filters.
+ */
+void avfilter_remove_lvpdnn(void);
+
```
These aren't used anymore - Will remove

jailuthra commented on doc/filters.texi:
```diff
@@ -14362,6 +14362,96 @@ lut2='if(lt(x,y),0,if(gt(x,y),pow(2,bdx)-1,pow(2,bdx-1))):if(lt(x,y),0,if(gt(x,y
 @end example
 @end itemize
 
+@section lvpdnn
```
Will update docs with new filter name `livepeer_dnn` and options

jailuthra commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -278,8 +288,18 @@ static DNNReturnType load_tf_model(TFModel *tf_model, const char *model_filename
     init_op = TF_GraphOperationByName(tf_model->graph, "init");
     sess_opts = TF_NewSessionOptions();
 
+    // FIXME: pass Livepeer filter config through the API
```
Still a pending TODO. Will try to use the `sess_config` API defined in the backend, populating it via a C/go utility function, as opposed to our `lp_sess_config` hardcoded protobuf values.

jailuthra commented on libavfilter/vf_livepeer_dnn.c:
```diff
@@ -0,0 +1,267 @@
+    struct AVFrame      *swscaleframe;  ///< Scaled image
+    FILE                *logfile;       ///< (Optional) Log classification probabilities in this file
+
+    int                 sample_rate;    ///< Run classification after every <sample_rate> frames
```
In LPMS I decided to use the `select` filter for this, as that prevents doing hwdownload of frames that we won't run detection on: https://github.com/livepeer/lpms/pull/246/files#diff-4ab68a4de67768f1acb5be217d4ec78bab27b99381025beedf82c11a5fa9c7aeR313

This wasn't a problem in our old filter because we manually copied frames from HW->SW as opposed to using `hwdownload`

yondonfu replied:
Neat use of the select filter!

IIUC the typical structure of a scene classification enabled DNN filtergraph will be select -> hwdownload -> livepeer_dnn such that only the frames outputted by the select filter are downloaded from GPU to CPU and then once those frames are in CPU memory they're passed to livepeer_dnn.

If this is the case, then can we remove all sample rate and frame counting logic in this filter and just assume that the user of this filter wants the model to use all frames passed in and a user can select the frames of interest prior to applying this filter?

jailuthra replied:
That sounds like a good idea - will remove!

jailuthra commented on libavfilter/vf_lvpdnn.c:
```diff
@@ -0,0 +1,545 @@
+/*
```
Old filter won't be needed anymore - Will remove once done with changes.

jailuthra reviewed (commented):
Made some self-review comments to help other reviewers:

yondonfu reviewed (commented):
Awesome to see how much less code we need for the livepeer_dnn filter since we can take advantage of the recent updates to the DNN backend APIs in ffmpeg!

jailuthra committed:
avfilter/vf_livepeer_dnn: Pre-execute model while initialization

This is required, as without it Tensorflow doesn't initialize
input/output tensors - and we end up spending that precious time at the
start of the first incoming frame.

jailuthra committed:
avfilter/dnn: Use backend APIs for sess_config

jailuthra committed:
avfilter,doc: Cleanup deadcode, add docs

oscar-davids reviewed (approved):
LGTM!

yondonfu commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -152,8 +152,8 @@ static DNNReturnType get_input_tf(void *model, DNNData *input, const char *input
-    // currently only NHWC is supported
-    av_assert0(dims[0] == 1);
+    //currently only NHWC is supported
+    av_assert0(dims[0] == 1 || dims[0] == -1);
```
I forget if this change would be useful upstream...But, in case it is we could isolate it in its own commit.

jailuthra replied:
Ahh just saw that somebody beat us to it! 

https://github.com/FFmpeg/FFmpeg/commit/1b1064054c8f3b4ad3b52d14f0c8ee1c4e8200fd

It's currently a part of `master`, so this change should automatically disappear from this commit probably whenever we rebase next and apply our commits on-top.

yondonfu reviewed (approved):
LGTM!

Just one comment about commit structure - feel free to consider and address during merge process.

jailuthra commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -152,8 +152,8 @@ static DNNReturnType get_input_tf(void *model, DNNData *input, const char *input
 
-    // currently only NHWC is supported
-    av_assert0(dims[0] == 1);
+    //currently only NHWC is supported
```
```suggestion
    // currently only NHWC is supported
```

jailuthra committed:
Update libavfilter/dnn/dnn_backend_tf.c
